spec/mem/tree.md

```diff
@@ -0,0 +1,564 @@
+    }
+}
+```
+
+We track each tag's permissions on specific memory locations (bytes)
+
+```rust
+pub enum AccessKind {
+    Read, 
+    Write,
+}
+
+pub enum Permission {
+    /// Represents a two-phase borrow during its reservation phase
+    Reserved, 
+    /// Represents an activated (written to) mutable reference
+    Active,
+    /// Represents a shared (immutable) reference
+    Frozen,
+    /// Represents a dead reference
+    Disabled, 
+}
+```
+
+We create one *tree* for tracking all pointers in each allocation.
+
+```rust
+pub struct TreeBorrowsAllocation {
+    /// The data stored in this allocation.
+    data: List<AbstractByte<TreeBorrowsProvenance>>,
+    /// The address where this allocation starts.
+    /// This is never 0, and `addr + data.len()` fits into a `usize`.
+    addr: Address,
+    /// The alignment that was requested for this allocation.
+    /// `addr` will be a multiple of this.
+    align: Align,
+    /// The kind of this allocation.
+    kind: AllocationKind,
+    /// Whether this allocation is still live.
+    live: bool,
+    /// The **TREE** for the Tree Borrows.
+    tree: Tree,
+}
+```
+
+The actual state machine for the permission transitions
+
+```rust
+impl TreeBorrowsAllocation {
+    fn child_read(permission: Permission) -> Result<Permission> {
+        match permission {
+            Permission::Reserved => ret(Permission::Reserved),
+            Permission::Active => ret(Permission::Active),
+            Permission::Frozen => ret(Permission::Frozen),
+            Permission::Disabled => throw_ub!("Tree Borrows: Child reading a pointer with the Disabled permission"),
+        }
+    }
+
+    fn child_write(permission: Permission) -> Result<Permission> {
+        match permission {
+            Permission::Reserved => ret(Permission::Active),
+            Permission::Active => ret(Permission::Active), 
+            Permission::Frozen => throw_ub!("Tree Borrows: Child writing a pointer with the Frozen permission"), 
+            Permission::Disabled => throw_ub!("Tree Borrows: Child writing a pointer with the Disabled permission"),
+        }
+    }
+
+    fn foreign_read(permission: Permission) -> Result<Permission> {
+        ret(
+            match permission {
+                Permission::Reserved => Permission::Reserved,
+                Permission::Active => Permission::Frozen,
+                Permission::Frozen => Permission::Frozen,
+                Permission::Disabled => Permission::Disabled,
+            }
+        )
+    }
+
+     fn foreign_write(permission: Permission) -> Result<Permission> {
+        ret(
+            match permission {
+                Permission::Reserved => Permission::Disabled,
+                Permission::Active => Permission::Disabled,
+                Permission::Frozen => Permission::Disabled,
+                Permission::Disabled => Permission::Disabled,
+            }
+        )
```

Comment:
```suggestion
        ret(Permission::Disabled)
```

Reply:
Actually this will have to change -- `Reserved` needs to track whether there is interior mutability, and then allow foreign writes. Also see https://github.com/rust-lang/miri/pull/3754 and [this test](https://github.com/rust-lang/miri/blob/05640c48b11508a33b471d5e00f0f8072d8dba58/tests/pass/tree_borrows/2phase-interiormut.rs). You should also add [all of these tests](https://github.com/rust-lang/miri/blob/6680b2fa09496b60c40c6ce09449f46efbf253d5/tests/pass/tree_borrows/tree-borrows.rs) (they might need some adjusting to work in minirust).

Reply:
To support this, I think we may need to modify the interface of `retag_ptr` to add an argument like `is_freeze`. Then implement a function like `is_trivially_freeze` in the https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_middle/ty/util.rs.html#1271. Do you think this makes sense? @RalfJung

Reply:
Hm... yeah we will need more information. Not just in `retag_ptr` but in `PtrType::Ref`.

Reply:
Thank you for the advice. I will send another small PR for this :)

Reply:
AFAIU, `UnsafeCell` is not currently supported. Maybe we should consider this after we support UnsafeCell?